File: apps/api/src/routes/projects.ts
import { Router } from "express";
import type { Response as ExpressResponse } from "express";
import { z } from "zod";
import { authMiddleware } from "../middleware/auth.js";
import { requirePermission } from "../middleware/permissions.js";
import { validate } from "../middleware/validate.js";
import {
  listProjects,
  createProject,
  updateProject,
  deleteProject,
  listProjectActivities,
  createProjectActivity,
  updateProjectActivity,
  deleteProjectActivity,
} from "../services/project-service.js";
import {
  listProjectActivityTasks,
  PROJECT_TASK_STATUSES,
  createProjectActivityTask,
  updateProjectActivityTask,
  deleteProjectActivityTask,
  reorderProjectActivityTasks,
} from "../services/project-task-service.js";

const router = Router();
router.use(authMiddleware);

type ProjectServiceErrorResult = {
  error: unknown;
  status?: number;
};

const respondProjectServiceError = (
  res: ExpressResponse,
  result: unknown,
  fallbackMessage: string,
): boolean => {
  if (typeof result !== "object" || result === null || !("error" in result)) {
    return false;
  }

  const errorResult = result as ProjectServiceErrorResult;
  res.status(typeof errorResult.status === "number" ? errorResult.status : 400).json({
    success: false,
    error: typeof errorResult.error === "string" ? errorResult.error : fallbackMessage,
  });
  return true;
};

const ProjectSchema = z.object({
  name: z.string().min(1).max(255),
  shortDescription: z.string().min(1).max(500),
  description: z.string().min(1),
  startDate: z.string().nullable().optional(),
  endDate: z.string().nullable().optional(),
  priority: z.enum(["low", "medium", "high", "urgent"]).default("medium"),
  status: z.enum(["active", "completed", "paused", "cancelled"]).default("active"),
});

const UpdateProjectSchema = ProjectSchema.extend({
  id: z.string().uuid(),
});

const ActivityBaseSchema = z.object({
  name: z.string().trim().min(1),
  description: z.string().trim().min(1),
  category: z.string().nullable().optional(),
  estimatedDays: z.number().nonnegative().nullable().optional(),
  startDate: z.string().nullable().optional(),
  endDate: z.string().nullable().optional(),
  priority: z.enum(["low", "medium", "high", "urgent"]).optional(),
  status: z.enum(["todo", "progress", "done", "blocked"]).optional(),
});

const UpdateActivitySchema = ActivityBaseSchema.extend({ id: z.string().uuid() });

const TaskOrderSchema = z.object({
  taskId: z.string().uuid(),
  status: z.enum(PROJECT_TASK_STATUSES),
  sort: z.number().int().nonnegative(),
});

const ReorderTasksSchema = z.object({
  tasksBeforeMove: z.array(TaskOrderSchema).min(1),
  tasksAfterMove: z.array(TaskOrderSchema).min(1),
});

const TaskBaseSchema = z.object({
  projectId: z.string().uuid(),
  projectActivityId: z.string().uuid(),
  name: z.string().trim().min(1),
  description: z.string().trim().min(1),
  category: z.string().nullable().optional(),
  estimatedDays: z.number().int().nonnegative().nullable().optional(),
  startDate: z.string().nullable().optional(),
  endDate: z.string().nullable().optional(),
  status: z.enum(PROJECT_TASK_STATUSES),
});

const CreateTaskSchema = TaskBaseSchema;

const UpdateTaskSchema = TaskBaseSchema.extend({
  id: z.string().uuid(),
});

const DeleteTaskSchema = z.object({
  id: z.string().uuid(),
});

const getQueryStringValue = (
  value: unknown,
): string | undefined =>
  typeof value === "string" && value.length > 0 ? value : undefined;

// ── Projects CRUD ──────────────────────────────────────────────────────────

router.get("/", requirePermission("projects", "view"), async (req, res) => {
  try {
    const search = getQueryStringValue(req.query.search);
    const status = getQueryStringValue(req.query.status);
    const priority = getQueryStringValue(req.query.priority);
    const result = await listProjects({ search, status, priority });
    if ("error" in result) {
      respondProjectServiceError(res, result, "Erro ao listar projetos.");
      return;
    }
    res.json({ success: true, data: result.data });
  } catch (err) {
    console.error("❌ [PROJECTS] GET:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.post("/", requirePermission("projects", "manage"), validate(ProjectSchema), async (req, res) => {
  try {
    const result = await createProject(req.body);
    if (respondProjectServiceError(res, result, "Erro ao criar projeto.")) {
      return;
    }
    res.status(201).json({ success: true, data: result.data, message: "Projeto criado com sucesso" });
  } catch (err) {
    console.error("❌ [PROJECTS] POST:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.put("/", requirePermission("projects", "manage"), validate(UpdateProjectSchema), async (req, res) => {
  try {
    const result = await updateProject(req.body);
    if ("error" in result) { respondProjectServiceError(res, result, "Erro ao atualizar projeto."); return; }
    res.json({ success: true, data: result.data, message: "Projeto atualizado com sucesso" });
  } catch (err) {
    console.error("❌ [PROJECTS] PUT:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.delete("/", requirePermission("projects", "manage"), async (req, res) => {
  const id = getQueryStringValue(req.query.id);
  if (!id) {
    res.status(400).json({ success: false, error: "ID do projeto é obrigatório." });
    return;
  }
  try {
    const result = await deleteProject(id);
    if ("error" in result) { respondProjectServiceError(res, result, "Erro ao excluir projeto."); return; }
    res.json({ success: true, message: "Projeto excluído com sucesso" });
  } catch (err) {
    console.error("❌ [PROJECTS] DELETE:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

// ── Project Activities ────────────────────────────────────────────────────

router.get("/:projectId/activities", requirePermission("projectActivities", "view"), async (req, res) => {
  try {
    const result = await listProjectActivities(String(req.params.projectId));
    if ("error" in result) { respondProjectServiceError(res, result, "Erro ao listar atividades."); return; }
    res.json({ success: true, data: { activities: result.data }, activities: result.data });
  } catch (err) {
    console.error("❌ [PROJECTS_ACTIVITIES] GET:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.post("/:projectId/activities", requirePermission("projectActivities", "manage"), validate(ActivityBaseSchema), async (req, res) => {
  try {
    const result = await createProjectActivity(String(req.params.projectId), req.body);
    if (result && "error" in result) { respondProjectServiceError(res, result, "Erro ao criar atividade."); return; }
    res.status(201).json({ success: true, data: { activity: result.data }, activity: result.data, message: "Atividade criada com sucesso" });
  } catch (err) {
    console.error("❌ [PROJECTS_ACTIVITIES] POST:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.put("/:projectId/activities", requirePermission("projectActivities", "manage"), validate(UpdateActivitySchema), async (req, res) => {
  try {
    const result = await updateProjectActivity(String(req.params.projectId), req.body);
    if ("error" in result) { respondProjectServiceError(res, result, "Erro ao atualizar atividade."); return; }
    res.json({ success: true, data: { activity: result.data }, activity: result.data, message: "Atividade atualizada com sucesso" });
  } catch (err) {
    console.error("❌ [PROJECTS_ACTIVITIES] PUT:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.delete("/:projectId/activities", requirePermission("projectActivities", "manage"), async (req, res) => {
  const activityId = getQueryStringValue(req.query.activityId);
  if (!activityId) {
    res.status(400).json({ success: false, error: "ID da atividade é obrigatório." });
    return;
  }
  try {
    const result = await deleteProjectActivity(String(req.params.projectId), activityId);
    if ("error" in result) { respondProjectServiceError(res, result, "Erro ao excluir atividade."); return; }
    res.json({ success: true, message: "Atividade excluída com sucesso" });
  } catch (err) {
    console.error("❌ [PROJECTS_ACTIVITIES] DELETE:", err);
    res.status(500).json({ success: false, error: "Erro interno do servidor" });
  }
});

router.get(
  "/:projectId/activities/:activityId/tasks",
  requirePermission("projectTasks", "view"),
  async (req, res) => {
    try {
      const result = await listProjectActivityTasks(
        String(req.params.projectId),
        String(req.params.activityId),
      );

      if ("error" in result) {
        respondProjectServiceError(res, result, "Erro ao listar tarefas da atividade.");
        return;
      }

      res.json({ success: true, data: result });
    } catch (err) {
      console.error("❌ [PROJECTS_ACTIVITY_TASKS] GET:", err);
      res.status(500).json({ success: false, error: "Erro interno do servidor" });
    }
  },
);

router.post(
  "/:projectId/activities/:activityId/tasks",
  requirePermission("projectTasks", "manage"),
  validate(CreateTaskSchema),
  async (req, res) => {
    try {
      const user = req.user!;
      const result = await createProjectActivityTask(
        String(req.params.projectId),
        String(req.params.activityId),
        user.id,
        req.body,
      );

      if ("error" in result) {
        respondProjectServiceError(res, result, "Erro ao criar tarefa.");
        return;
      }

      res.status(201).json({
        success: true,
        data: { task: result.task },
        task: result.task,
        message: "Tarefa criada com sucesso",
      });
    } catch (err) {
      console.error("❌ [PROJECTS_ACTIVITY_TASKS] POST:", err);
      res.status(500).json({ success: false, error: "Erro interno do servidor" });
    }
  },
);

router.put(
  "/:projectId/activities/:activityId/tasks",
  requirePermission("projectTasks", "manage"),
  validate(UpdateTaskSchema),
  async (req, res) => {
    try {
      const user = req.user!;
      const result = await updateProjectActivityTask(
        String(req.params.projectId),
        String(req.params.activityId),
        user.id,
        req.body,
      );

      if ("error" in result) {
        respondProjectServiceError(res, result, "Erro ao atualizar tarefa.");
        return;
      }

      res.json({
        success: true,
        data: { task: result.task },
        task: result.task,
        message: "Tarefa atualizada com sucesso",
      });
    } catch (err) {
      console.error("❌ [PROJECTS_ACTIVITY_TASKS] PUT:", err);
      res.status(500).json({ success: false, error: "Erro interno do servidor" });
    }
  },
);

router.delete(
  "/:projectId/activities/:activityId/tasks",
  requirePermission("projectTasks", "manage"),
  validate(DeleteTaskSchema),
  async (req, res) => {
    try {
      const result = await deleteProjectActivityTask(
        String(req.params.projectId),
        String(req.params.activityId),
        req.body.id,
      );

      if ("error" in result) {
        respondProjectServiceError(res, result, "Erro ao excluir tarefa.");
        return;
      }

      res.json({ success: true, message: "Tarefa excluída com sucesso" });
    } catch (err) {
      console.error("❌ [PROJECTS_ACTIVITY_TASKS] DELETE:", err);
      res.status(500).json({ success: false, error: "Erro interno do servidor" });
    }
  },
);

router.patch(
  "/:projectId/activities/:activityId/tasks",
  requirePermission("projectTasks", "manage"),
  validate(ReorderTasksSchema),
  async (req, res) => {
    try {
      const user = req.user!;
      const result = await reorderProjectActivityTasks(
        String(req.params.projectId),
        String(req.params.activityId),
        user.id,
        req.body.tasksBeforeMove,
        req.body.tasksAfterMove,
      );

      if ("error" in result) {
        if (result.status === 409) {
          const conflictTasks = result.tasks ?? [];

          res.status(409).json({
            success: false,
            error: result.error,
            data: { tasks: conflictTasks },
            tasks: conflictTasks,
          });
          return;
        }

        respondProjectServiceError(res, result, "Erro ao reordenar tarefas.");
        return;
      }

      res.json({
        success: true,
        data: { tasks: result.tasks },
        tasks: result.tasks,
        message: "Movimentação salva com sucesso",
      });
    } catch (err) {
      console.error("❌ [PROJECTS_ACTIVITY_TASKS] PATCH:", err);
      res.status(500).json({ success: false, error: "Erro interno do servidor" });
    }
  },
);

export default router;

File: apps/backend/src/silo/api/routers/projects.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.engine import Connection

from silo.api.dependencies import CurrentUser, get_db, require_permission
from silo.api.responses import build_success_payload, json_error_response
from silo.services.common import is_service_error, service_error_response, service_failure
from silo.services.project_portal import (
    PROJECT_TASK_STATUSES,
    create_project,
    create_project_activity,
    create_project_activity_task,
    delete_project,
    delete_project_activity,
    delete_project_activity_task,
    list_project_activities,
    list_project_activity_tasks,
    list_projects,
    reorder_project_activity_tasks,
    update_project,
    update_project_activity,
    update_project_activity_task,
)
from silo.storage.uploads import delete_upload_file, is_safe_filename, list_upload_files

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.get("")
@router.get("/")
async def get_projects(
    search: str | None = Query(default=None),
    status: str | None = Query(default=None),
    priority: str | None = Query(default=None),
    _current_user: object = Depends(require_permission("projects", "view")),
    db: Connection = Depends(get_db),
):
    result = list_projects(db, search=search, status=status, priority=priority)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao listar projetos.")
        assert response is not None
        return response
    return build_success_payload(result["data"])


@router.post("")
@router.post("/")
async def post_project(
    payload: dict[str, object],
    _current_user: object = Depends(require_permission("projects", "manage")),
    db: Connection = Depends(get_db),
):
    result = create_project(db, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao criar projeto.")
        assert response is not None
        return response
    return JSONResponse(
        status_code=201,
        content=build_success_payload(result["data"], message="Projeto criado com sucesso"),
    )


@router.put("")
@router.put("/")
async def put_project(
    payload: dict[str, object],
    _current_user: object = Depends(require_permission("projects", "manage")),
    db: Connection = Depends(get_db),
):
    result = update_project(db, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao atualizar projeto.")
        assert response is not None
        return response
    return build_success_payload(result["data"], message="Projeto atualizado com sucesso")


@router.delete("")
@router.delete("/")
async def delete_project_route(
    id: str | None = Query(default=None),
    _current_user: object = Depends(require_permission("projects", "manage")),
    db: Connection = Depends(get_db),
):
    if not id:
        return json_error_response(400, "ID do projeto é obrigatório.")

    result = delete_project(db, id)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao excluir projeto.")
        assert response is not None
        return response
    return build_success_payload(message="Projeto excluído com sucesso")


@router.get("/images")
async def list_project_images(
    _current_user: object = Depends(require_permission("projects", "view")),
):
    items = list_upload_files("projects")
    return build_success_payload({"items": items})


@router.delete("/images")
async def delete_project_image(
    filename: str | None = Query(default=None),
    _current_user: object = Depends(require_permission("projects", "manage")),
):
    if not filename or not is_safe_filename(filename):
        return service_error_response(
            service_failure("Nome de arquivo inválido", 400),
            "Erro ao excluir imagem",
        )

    delete_upload_file("projects", filename)
    return build_success_payload(message="Imagem excluída com sucesso")


@router.get("/{projectId}/activities")
async def get_project_activities(
    projectId: str,
    _current_user: object = Depends(require_permission("projectActivities", "view")),
    db: Connection = Depends(get_db),
):
    result = list_project_activities(db, projectId)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao listar atividades.")
        assert response is not None
        return response

    activities = result["data"]["activities"]
    return JSONResponse(
        content={
            "success": True,
            "data": {"activities": activities},
            "activities": activities,
        }
    )


@router.post("/{projectId}/activities")
async def post_project_activity(
    projectId: str,
    payload: dict[str, object],
    _current_user: object = Depends(require_permission("projectActivities", "manage")),
    db: Connection = Depends(get_db),
):
    result = create_project_activity(db, projectId, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao criar atividade.")
        assert response is not None
        return response

    activity = result["data"]["activity"]
    return JSONResponse(
        status_code=201,
        content={
            "success": True,
            "data": {"activity": activity},
            "activity": activity,
            "message": "Atividade criada com sucesso",
        },
    )


@router.put("/{projectId}/activities")
async def put_project_activity(
    projectId: str,
    payload: dict[str, object],
    _current_user: object = Depends(require_permission("projectActivities", "manage")),
    db: Connection = Depends(get_db),
):
    result = update_project_activity(db, projectId, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao atualizar atividade.")
        assert response is not None
        return response

    activity = result["data"]["activity"]
    return JSONResponse(
        content={
            "success": True,
            "data": {"activity": activity},
            "activity": activity,
            "message": "Atividade atualizada com sucesso",
        },
    )


@router.delete("/{projectId}/activities")
async def delete_project_activity_route(
    projectId: str,
    activityId: str | None = Query(default=None),
    _current_user: object = Depends(require_permission("projectActivities", "manage")),
    db: Connection = Depends(get_db),
):
    if not activityId:
        return json_error_response(400, "ID da atividade é obrigatório.")

    result = delete_project_activity(db, projectId, activityId)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao excluir atividade.")
        assert response is not None
        return response
    return build_success_payload(message="Atividade excluída com sucesso")


@router.get("/{projectId}/activities/{activityId}/tasks")
async def get_project_activity_tasks(
    projectId: str,
    activityId: str,
    _current_user: object = Depends(require_permission("projectTasks", "view")),
    db: Connection = Depends(get_db),
):
    result = list_project_activity_tasks(db, projectId, activityId)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao listar tarefas da atividade.")
        assert response is not None
        return response
    return build_success_payload(result["data"])


@router.post("/{projectId}/activities/{activityId}/tasks")
async def post_project_activity_task(
    projectId: str,
    activityId: str,
    payload: dict[str, object],
    current_user: CurrentUser = Depends(require_permission("projectTasks", "manage")),
    db: Connection = Depends(get_db),
):
    result = create_project_activity_task(db, projectId, activityId, current_user.id, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao criar tarefa.")
        assert response is not None
        return response

    task = result["data"]["task"]
    return JSONResponse(
        status_code=201,
        content={
            "success": True,
            "data": {"task": task},
            "task": task,
            "message": "Tarefa criada com sucesso",
        },
    )


@router.put("/{projectId}/activities/{activityId}/tasks")
async def put_project_activity_task(
    projectId: str,
    activityId: str,
    payload: dict[str, object],
    current_user: CurrentUser = Depends(require_permission("projectTasks", "manage")),
    db: Connection = Depends(get_db),
):
    result = update_project_activity_task(db, projectId, activityId, current_user.id, payload)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao atualizar tarefa.")
        assert response is not None
        return response

    task = result["data"]["task"]
    return JSONResponse(
        content={
            "success": True,
            "data": {"task": task},
            "task": task,
            "message": "Tarefa atualizada com sucesso",
        },
    )


@router.delete("/{projectId}/activities/{activityId}/tasks")
async def delete_project_activity_task_route(
    projectId: str,
    activityId: str,
    payload: dict[str, object],
    _current_user: object = Depends(require_permission("projectTasks", "manage")),
    db: Connection = Depends(get_db),
):
    task_id = _optional_str(payload.get("id"))
    if not task_id:
        return json_error_response(400, "ID é obrigatório.")

    result = delete_project_activity_task(db, projectId, activityId, task_id)
    if is_service_error(result):
        response = service_error_response(result, "Erro ao excluir tarefa.")
        assert response is not None
        return response
    return build_success_payload(message="Tarefa excluída com sucesso")


@router.patch("/{projectId}/activities/{activityId}/tasks")
async def patch_project_activity_tasks(
    projectId: str,
    activityId: str,
    payload: dict[str, object],
    current_user: CurrentUser = Depends(require_permission("projectTasks", "manage")),
    db: Connection = Depends(get_db),
):
    before = payload.get("tasksBeforeMove")
    after = payload.get("tasksAfterMove")
    if not isinstance(before, list) or not isinstance(after, list):
        return json_error_response(400, "Dados de movimentação inválidos.")

    result = reorder_project_activity_tasks(db, projectId, activityId, current_user.id, before, after)
    if is_service_error(result):
        status = int(result.get("status") or 400)
        if status == 409:
            conflict_tasks = []
            data = result.get("data")
            if isinstance(data, dict) and isinstance(data.get("tasks"), list):
                conflict_tasks = data["tasks"]
            return JSONResponse(
                status_code=409,
                content={
                    "success": False,
                    "error": "KANBAN_OUTDATED",
                    "data": {"tasks": conflict_tasks},
                    "tasks": conflict_tasks,
                },
            )

        response = service_error_response(result, "Erro ao reordenar tarefas.")
        assert response is not None
        return response

    tasks = result["data"]["tasks"]
    return JSONResponse(
        content={
            "success": True,
            "data": {"tasks": tasks},
            "tasks": tasks,
            "message": "Movimentação salva com sucesso",
        },
    )


def _optional_str(value: object | None) -> str | None:
    return value if isinstance(value, str) else None
